base/compiler/abstractinterpretation.jl

```diff
@@ -189,16 +189,18 @@ function abstract_call_method(method::Method, @nospecialize(sig), sparams::Simpl
     cyclei = 0
     infstate = sv
     edgecycle = false
+    checked_method = method_for_specialization_heuristics(method, sig, sparams, sv.params.world)
     while !(infstate === nothing)
         infstate = infstate::InferenceState
-        if method === infstate.linfo.def
-            if infstate.linfo.specTypes == sig
-                # avoid widening when detecting self-recursion
-                # TODO: merge call cycle and return right away
-                topmost = nothing
-                edgecycle = true
-                break
-            end
+        if infstate.linfo.specTypes == sig
```

Comment:
Should checks against signatures be replaced in the same manner as checks against methods?